### PR TITLE
Replace signed array access indices in unsiged prange loop body

### LINKED ISSUE #2846: ParallelAccelerator regression due to vectorization changes
We changed parfor indices to `uintp` in #2727, which causes type mismatch with programs index variables (coming from `prange` or `internal_prange`). This has a cascading effect in the optimization pipeline.

The simple example below demonstrates the issue. The array A should eventually be eliminated, but `i.23` is not replaced with `parfor_index.32` in copy propagation and so on.

```python
@numba.njit(parallel=True)
def test_impl(n):
    A = np.ones(n)
    return A.sum()
```

Optimized IR without the index type change:
```
label 0:
    n = arg(0, name=n)                       ['n']
---begin parfor 0---
index_var =  parfor_index.32
LoopNest(index_variable = parfor_index.32, range = (0, n, 1)) 
init block:
    $val.7.60 = const(float64, 0.0) 
    val.7 = $val.7.60 
label 52: 
    $expr_out_var.36 = const(float64, 1.0) 
    $val.7.59 = inplace_binop(fn=+=, immutable_fn=+, lhs=val.7, rhs=$expr_out_var.36, static_lhs=<object object at 0x7f8d7cf96e00>, static_rhs=<object object at 0x7f8d7cf96e00>)
    val.7 = $val.7.59                        ['$val.7.59', 'val.7']
----end parfor 0----
    $0.8 = cast(value=val.7)                 ['$0.8', 'val.7']
    return $0.8                              ['$0.8']
```

Optimized IR with the index type change:
```
label 0:
    n = arg(0, name=n)                       ['n']
---begin parfor 0---
index_var =  parfor_index.32
LoopNest(index_variable = parfor_index.32, range = (0, n, 1))
init block:
    $np_g_var.33 = global(np: <module 'numpy' from '/home/etotoni/pse-hpc/python/anaconda3_5.01/envs/PA/lib/python3.6/site-packages/numpy/__init__.py'>) ['$np_g_var.33']
    $empty_attr_attr.34 = getattr(value=$np_g_var.33, attr=empty) ['$empty_attr_attr.34', '$np_g_var.33']
    $np_typ_var.35 = getattr(value=$np_g_var.33, attr=float64) ['$np_g_var.33', '$np_typ_var.35']
    $A.38 = call $empty_attr_attr.34(n, $np_typ_var.35, func=$empty_attr_attr.34, args=[Var(n, tmp_tests/reduce_scalar.py (6)), Var($np_typ_var.35, tmp_tests/reduce_scalar.py (6))], kws=(), vararg=None) ['$A.38', '$empty_attr_attr.34', '$np_typ_var.35', 'n']
    $val.7.60 = const(float64, 0.0)          ['$val.7.60']
    val.7 = $val.7.60                        ['$val.7.60', 'val.7']
label 52:
    $expr_out_var.36 = const(float64, 1.0)   ['$expr_out_var.36']
    $A.38[parfor_index.32] = $expr_out_var.36 ['$A.38', '$expr_out_var.36', 'parfor_index.32']
    i.23 = parfor_index.32                   ['i.23', 'parfor_index.32']
    $34.5.27 = getitem(value=$A.38, index=i.23) ['$34.5.27', '$A.38', 'i.23']
    $val.7.59 = inplace_binop(fn=+=, immutable_fn=+, lhs=val.7, rhs=$34.5.27, static_lhs=<object object at 0x7fcb9b8ce6e0>, static_rhs=<object object at 0x7fcb9b8ce6e0>) ['$34.5.27', '$val.7.59', 'val.7']
    val.7 = $val.7.59                        ['$val.7.59', 'val.7']
----end parfor 0---- 
    $0.8 = cast(value=val.7)                 ['$0.8', 'val.7']
    return $0.8                              ['$0.8']

```

### PR DESCRIPTION
**Pull request by ehsantn**
Closes #2846. Replaces array access indices if prange index is changed to be unsigned in translation. This helps downstream optimizations. See #2846 for explanation and example.

### CODE REVIEW

File: numba/parfor.py [hunk 1]
```diff
@@ -1009,6 +1022,83 @@ def find_mask_from_size(size_var):
...
+        Replace array access indices in a loop body with a new index.
+        index_set has all the variables that are equivalent to loop index.
+        """
+        # treat new index like others since replacing it with itself is ok
+        index_set.add(new_index.name)
+
+        # add dummy return to last block for CFG computation
+        # max is last block since we add it manually for prange
+        last_label = max(loop_body.keys())
+        loop_body[last_label].body.append(
```
Comment (DrTodd13) [inline numba/parfor.py]: Maybe we should consider doing this add/remove dummy return in Python "with" style.

File: numba/parfor.py [hunk 2]
```diff
@@ -1009,6 +1022,83 @@ def find_mask_from_size(size_var):
...
+            ir.Return(0, ir.Loc("parfors_dummy", -1)))
+
+        labels = find_topo_order(loop_body)
+
+        # remove dummy return
+        loop_body[last_label].body.pop()
```
Comment (DrTodd13) [inline numba/parfor.py]: Should be the end of the "with" block and therefore hidden from the programmer.

File: numba/parfor.py [hunk 3]
```diff
@@ -1009,6 +1022,83 @@ def find_mask_from_size(size_var):
...
+
+        first_label = labels[0]
+        added_indices = set()
+
+        for l in labels:
```
Comment (DrTodd13) [inline numba/parfor.py]: Documentation before this block saying what it is doing.

File: numba/parfor.py [hunk 4]
```diff
@@ -2469,32 +2559,56 @@ def remove_dead_parfor(parfor, lives, arg_aliases, alias_map, typemap):
     """ remove dead code inside parfor including get/sets
     """
 
+    # add dummy return to last block for CFG computation
+    # max is last block since we add it manually for parfors
+    last_label = max(parfor.loop_body.keys())
+    parfor.loop_body[last_label].body.append(
```
Comment (DrTodd13) [inline numba/parfor.py]: Same conversion to "with" style I suggest above.

File: numba/parfor.py [hunk 5]
```diff
@@ -2469,32 +2559,56 @@ def remove_dead_parfor(parfor, lives, arg_aliases, alias_map, typemap):
...
+        ir.Return(0, ir.Loc("parfors_dummy", -1)))
+
+    labels = find_topo_order(parfor.loop_body)
+
+    # remove dummy return
+    parfor.loop_body[last_label].body.pop()
```
Comment (DrTodd13) [inline numba/parfor.py]: End of suggested "with" block.

File: numba/parfor.py [hunk 6]
```diff
@@ -2887,6 +3032,18 @@ def parfor_typeinfer(parfor, typeinferer):
 
 typeinfer.typeinfer_extensions[Parfor] = parfor_typeinfer
 
+def build_parfor_definitions(parfor, definitions=None):
+    """get variable definition table for parfors"""
+    if definitions is None:
+        definitions = dict()
+
+    blocks = wrap_parfor_blocks(parfor)
```
Comment (DrTodd13) [inline numba/parfor.py]: This wrap/unwrap pattern should also be a "with" style probably.

File: numba/parfor.py (hunk 1 above)
Comment (ehsantn) [reply numba/parfor.py]: Good idea. Done.

File: numba/parfor.py (hunk 2 above)
Comment (ehsantn) [reply numba/parfor.py]: done.

File: numba/parfor.py (hunk 3 above)
Comment (ehsantn) [reply numba/parfor.py]: done.

File: numba/parfor.py (hunk 4 above)
Comment (ehsantn) [reply numba/parfor.py]: done.

File: numba/parfor.py (hunk 5 above)
Comment (ehsantn) [reply numba/parfor.py]: done.

File: numba/parfor.py (hunk 6 above)
Comment (ehsantn) [reply numba/parfor.py]: I agree. There are subtle differences in its usage so it needs more major refactoring. Let's leave it for another PR.

**Comment by ehsantn** [comment]
@DrTodd13 Thanks. I applied the changes. I agree we should improve documentation.

**Pull request merged**